BinaryEncoding.md

```diff
@@ -147,28 +147,32 @@ A module may contain at most one import table section.
 | func_name | `uint32` | offset from the start of the module of the string representing the function name |
 
 
-### Functions section
+### Function Signatures section
 
-ID: `functions`
+ID: `function_signatures`
 
-The Functions section declares the functions in the module and must be preceded by a [Signatures](#signatures-section) section. A module may contain at most one functions section.
+The Function Signatures section declares the signatures of all functions in the
+module and must be preceded by the [Signatures](#signatures-section) section. A
+module may contain at most one functions signatures section.
 
 | Field | Type | Description |
 | ----- |  ----- | ----- |
-| count | `varuint32` | count of function entries to follow |
-| entries | `function_entry*` | repeated function entries as described below |
+| count | `varuint32` | count of signature indices to follow |
+| signatures | `varuint32*` | sequence of indices into the Signature section |
 
-#### Function Entry
+### Function Bodies section
 
-Each function entry describes a function that can be optionally named, imported and/or exported. Non-imported functions
-must contain a function body. Imported and exported functions must have a name. Imported functions do not contain a body.
+ID: `function_bodies`
 
-| Field | Type |  Present?  | Description |
+The Function Bodies section assigns a body to every function in the module and
+must be preceded by the [Function Signatures](#function-signatures-section) section.
+The count of function signatures and function bodies must be the same and the `i`th
+signature corresponds to the `i`th function body.
+
+| Field | Type |  Description |
 | ----- |  ----- |  ----- |  ----- |
-| flags | `uint8` | always | flags indicating attributes of a function <br>bit `1` : the function is an import<br>bit `2` : the function has local variables<br>bit `3` : the function is an export |
-| signature | `uint16` | always | index into the Signature section |
-| body size | `uint16` | `flags[0] == 0` | size of function body to follow, in bytes |
```

Comment:
On Mon, Mar 7, 2016 at 2:18 PM, Luke Wagner <notifications@github.com>
wrote:

> In BinaryEncoding.md
> <https://github.com/WebAssembly/design/pull/591#discussion_r55282783>:
>
> >  | ----- |  ----- |  ----- |  ----- |
> > -| flags | `uint8` | always | flags indicating attributes of a function <br>bit `1` : the function is an import<br>bit `2` : the function has local variables<br>bit `3` : the function is an export |
> > -| signature | `uint16` | always | index into the Signature section |
> > -| body size | `uint16` | `flags[0] == 0` | size of function body to follow, in bytes |
>
> The existing body byte size is sufficient to know when the body is done.
> @rossberg-chromium <https://github.com/rossberg-chromium> was suggesting
> to add a body terminator. This PR reflects what's in SM which I had thought
> is what's in V8 as well. Perhaps we could add the body terminator in a
> separate PR?
>
The byte terminator seems somewhat redundant to me, and even in-person
discussions with @rossberg-chromium haven't convinced me otherwise yet :-)


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/591/files#r55282783>.
>